## Instruction

> 1. Rename Assignment-01-###.ipynb where ### is your student ID.
> 2. The deadline of Assignment-01 is 23:59pm, 03-31-2024
>
> 3. In this assignment, you will
>    1) explore Wikipedia text data
>    2) build language models
>    3) build NB and LR classifiers

## Task0 - Download datasets
> Download the preprocessed data, enwiki-train.json and enwiki-test.json from the Assignment-01 folder. In the data file, each line contains a Wikipedia page with attributes, title, label, and text. There are 1000 records in the train file and 100 records in test file with ten categories.

## Task1 - Data exploring and preprocessing

> 1) Print out how many documents are in each class  (for both train and test dataset)

代码参考demo code L01以及L02 nltk 数据包的学习

In [5]:
import json
from collections import Counter
with open('enwiki-train.json', 'r') as file:
    train_data = file.read()

json_objects_train = train_data.strip().split('\n')

# 提取所有class
classes_train = []
for obj in json_objects_train:
    json_data = json.loads(obj)
    classes = json_data.get('label')
    if classes:
        classes_train.append(classes)

train_classes_count = Counter(classes_train)
print("Count of documents in each class for train dataset:")
for classes, count in train_classes_count.items():
    print(classes, ":", count)
    
with open('enwiki-test.json', 'r') as file: 
    test_data = file.read()

json_objects_test = test_data.strip().split('\n')

# 提取所有class
classes_test = []
for obj in json_objects_test:
    json_data = json.loads(obj)
    classes = json_data.get('label')
    if classes:
        classes_test.append(classes)

test_classes_count = Counter(classes_test)

print("Count of documents in each class for test dataset:")
for classes, count in test_classes_count.items():
    print(classes, ":", count)


Count of documents in each class for train dataset:
Film : 100
Book : 100
Politician : 100
Writer : 100
Food : 100
Actor : 70
Animal : 80
Software : 130
Artist : 100
Disease : 120
Count of documents in each class for test dataset:
Film : 10
Book : 10
Politician : 10
Writer : 10
Food : 10
Actor : 10
Animal : 10
Software : 10
Artist : 10
Disease : 10


> 2) Print out the average number of sentences in each class.
>    You may need to use sentence tokenization of NLTK.
>    (for both train and test dataset)


In [8]:
# Your code goes to here
import nltk
## train
def count_sent_average_train(label, dataset):
    average_num = 0
    docu_num = train_classes_count.get(label)
    for i in range(len(dataset)):
        json_data = json.loads(dataset[i])
        classes = json_data.get('label')
        text = json_data.get('text')
        if classes == label:
            sentence = nltk.sent_tokenize(text)
            average_num += len(sentence) / docu_num
            
    return average_num
print("In train data:")
for i in set(classes_train):
    print('average number of sentenses in', i, 'is: {}'.format(count_sent_average_train(i, json_objects_train)))

## test
def count_sent_average_test(label, dataset):
    average_num = 0
    docu_num = test_classes_count.get(label)
    for i in range(len(dataset)):
        json_data = json.loads(dataset[i])
        classes = json_data.get('label')
        text = json_data.get('text')
        if classes == label:
            sentence = nltk.sent_tokenize(text)
            average_num += len(sentence) / docu_num
            
    return average_num
print("In test data:")
for i in set(classes_test):
    print('average number of sentenses in', i, 'is: {}'.format(count_sent_average_test(i, json_objects_test)))

In train data:
average number of sentenses in Food is: 175.24000000000004
average number of sentenses in Animal is: 70.375
average number of sentenses in Film is: 438.56000000000023
average number of sentenses in Software is: 260.9538461538461
average number of sentenses in Writer is: 420.32000000000016
average number of sentenses in Book is: 400.36000000000024
average number of sentenses in Politician is: 706.1999999999999
average number of sentenses in Artist is: 306.46999999999997
average number of sentenses in Actor is: 76.69999999999997
average number of sentenses in Disease is: 404.90000000000015
In test data:
average number of sentenses in Food is: 107.60000000000001
average number of sentenses in Animal is: 46.800000000000004
average number of sentenses in Film is: 364.70000000000005
average number of sentenses in Software is: 160.1
average number of sentenses in Writer is: 294.90000000000003
average number of sentenses in Book is: 295.90000000000003
average number of sentenses

> 3) Print out the average number of tokens in each class
>    (for both train and test dataset)

In [64]:
# Your code goes to here
## train
def count_token_average_train(label, dataset):
    average_num = 0
    docu_num = train_classes_count.get(label)
    for i in range(len(dataset)):
        json_data = json.loads(dataset[i])
        classes = json_data.get('label')
        text = json_data.get("text")
        if classes == label:
            token = nltk.word_tokenize(text)
            average_num += len(token) / docu_num
    return average_num

print("In train data:")
for i in set(classes_train):
    print('average number of tokens in', i, 'is: {}'.format(count_token_average_train(i, json_objects_train)))


## test
def count_token_average_test(label, dataset):
    average_num = 0
    docu_num = test_classes_count.get(label)
    for i in range(len(dataset)):
        json_data = json.loads(dataset[i])
        classes = json_data.get('label')
        text = json_data.get('text')
        if classes == label:
            token = nltk.word_tokenize(text)
            average_num += len(token) / docu_num
            
    return average_num

print("In test data:")
for i in set(classes_test):
    print('average number of tokens in', i, 'is: {}'.format(count_token_average_test(i, json_objects_test)))

In train data:
average number of tokens in Film is: 11895.279999999999
average number of tokens in Actor is: 1868.8428571428574
average number of tokens in Writer is: 11849.91
average number of tokens in Disease is: 9322.958333333332
average number of tokens in Food is: 3904.149999999999
average number of tokens in Animal is: 1521.9250000000002
average number of tokens in Software is: 6302.300000000002
average number of tokens in Politician is: 18644.3
average number of tokens in Artist is: 8212.909999999996
average number of tokens in Book is: 10540.509999999998
In test data:
average number of tokens in Film is: 9292.9
average number of tokens in Actor is: 677.5
average number of tokens in Writer is: 8499.4
average number of tokens in Disease is: 6988.8
average number of tokens in Food is: 2445.5
average number of tokens in Animal is: 885.6
average number of tokens in Software is: 3972.7999999999993
average number of tokens in Politician is: 15204.300000000003
average number of tokens

> 4) For each sentence in the document, remove punctuations and other special characters so that each sentence only contains English words and numbers. To make your life easier, you can make all words as lower cases. For each class, print out the first article's name and the processed first 40 words. (for both train and test dataset)

In [80]:
# Your code goes to here
import nltk
def remove_punc_and_others(dataset):
    pattern = r'[a-zA-Z-9]+'
    for i in range(len(dataset)):
        json_data = json.loads(dataset[i])
        text = json_data.get("text")
        words = nltk.regexp_tokenize(text, pattern)
        cleaned_text = ' '.join(word for word in words)
        print(cleaned_text)

## train       
def print_fun(label, dataset):
    pattern = r'[a-zA-Z-9]+'
    for i in range(len(dataset)):
        json_data = json.loads(dataset[i])
        classes = json_data.get('label')
        text = json_data.get("text")
        title = json_data.get("title")
        if classes == label:
            words = nltk.regexp_tokenize(text, pattern)
            cleaned_text = ' '.join(word for word in words[:40])
            print(title,':', cleaned_text)
            break


print("In train dataset:")
for i in set(classes_train):
    print_fun(i, json_objects_train)

## test
print("In test dataset:")
for i in set(classes_test):
    print_fun(i, json_objects_test)


In train dataset:
Citizen_Kane : Citizen Kane is a 9 American drama film produced by directed by and starring Orson Welles He also co-wrote the screenplay with Herman J Mankiewicz The picture was Welles first feature film Citizen Kane is considered by many critics and
Roman_Polanski : Roman Polanski born Raymond Thierry Liebling on August 9 is a Polish French film director producer screenwriter and actor His Polish Jewish parents moved the family from Paris back to Krak w in 9 Two years later Poland was invaded
Mircea_Eliade : Mircea Eliade April 9 was a Romanian historian of religion fiction writer philosopher and professor at the University of Chicago He was a leading interpreter of religious experience who established paradigms in religious studies that persist to this day His
Domestic_violence : Domestic violence also called domestic abuse or family violence is violence or other abuse that occurs in a domestic setting such as in a marriage or cohabitation Domestic violence is often

## Task2 - Build language models

> 1) Based on the training dataset, build unigram, bigram, and trigram language models using Add-one smoothing technique. It is encouraged to implement models by yourself. If you use public code, please cite it.


在理解清楚https://github.com/joshualoehr/ngram-language-model 的 （即老师在demo code L02里提供的）代码的前提下，按照他大致的整体结构编写而成

In [56]:
# Your code goes to here
## 代码的整体结构学习 https://github.com/joshualoehr/ngram-language-model 的
import nltk
from itertools import product
import math

def token_text(dataset, n, k):
    sos = "<s> " * (n-1) if n > 1 else "<s> "
    tokenized_word = []
    for i in range(len(dataset)):
        obj = json.loads(dataset[i])
        text = obj.get("text")
        text = text.replace("\n", " ")
        text_split = nltk.sent_tokenize(text)
        tokenized_sentences = ['{}{} {}'.format(sos, sent, "</s>").split() for sent in text_split]
        tokens = [token for sublist in tokenized_sentences for token in sublist]
        tokenized_word.extend(tokens)
    vocab = nltk.FreqDist(tokenized_word)
    tokenized_word = [token if vocab[token] > k else "<UNK>" for token in tokenized_word]
    return tokenized_word, tokenized_sentences

In [13]:
class LanguageModel():
    def __init__(self, tokens, n):
        self.tokens = tokens
        self.n = n
        self.vocab  = nltk.FreqDist(self.tokens)
        self.vocabulary_size = len(set(self.tokens)) 
        self.model = self.train_model()
        self.mask = list(reversed(list(product((0,1), repeat=n))))
    def train_model(self):
        """建立模型
        
        Returns:
            N-gram概率
        """
        if self.n == 1:
            num_tokens = len(self.tokens)
            # n_grams = nltk.ngrams(self.tokens, self.n)
            return { (unigram,): (count + 1) /(num_tokens + 1 * self.vocabulary_size) for unigram, count in self.vocab.items() }
        else:
            n_grams = nltk.ngrams(self.tokens, self.n)
            n_vocab = nltk.FreqDist(n_grams)

            m_grams = nltk.ngrams(self.tokens, self.n-1)
            m_vocab = nltk.FreqDist(m_grams)
            def smoothed_count(n_gram, n_count):
                m_gram = n_gram[:-1]
                m_count = m_vocab[m_gram]
                return (n_count + 1) / (m_count + 1 * self.vocabulary_size)
            return { n_gram: smoothed_count(n_gram, count) for n_gram, count in n_vocab.items() }

    def change_testtoken(self, ngram):
        """_summary_
            没有出现的字词换为<NUK>
        """
        mask = lambda ngram, bitmask: tuple((token if flag == 1 else "<UNK>" for token,flag in zip(ngram, bitmask)))

        ngram = (ngram,) if type(ngram) is str else ngram
        for possible_known in [mask(ngram, bitmask) for bitmask in self.mask]:
            if possible_known in self.model:
                return possible_known

            
    def perplexity(self, test_tokens):
        test_ngrams = nltk.ngrams(test_tokens, self.n)
        N = len(test_tokens)
        known_ngrams  = (self.change_testtoken(ngram) for ngram in test_ngrams)
        probabilities = [self.model[ngram] for ngram in known_ngrams]

        return math.exp((-1/N) * sum(map(math.log, probabilities)))
    
    def generate_sentences(self, num, min_len = 12, max_len = 24):
        def max_prob_word(prev, i, without=[]):
            blacklist = ["<UNK>"] + without
            candidates = ((ngram[-1],prob) for ngram, prob in self.model.items() if ngram[:-1]== prev)
            candidates = [candidate for candidate in candidates if candidate[0] not in blacklist]
            word_list = sorted(candidates, key=lambda candidate: candidate[1], reverse=True)
            if len(word_list) == 0:
                return ("</s>", 1)
            elif prev != () and prev[-1] != "<s>":
                return word_list[0]
        
        for i in range(num):
            sent, prob = ["<s>"] * max(1, self.n-1), 1
            while sent[-1] != "</s>":
                if self.n == 1:
                    prev = () 
                else:
                    prev = tuple(sent[-(self.n-1):])
                    
                blacklist = sent + (["</s>"] if len(sent) < min_len else [])
                next_token, next_prob = max_prob_word(prev, i, without=blacklist)
                sent.append(next_token)
                prob *= next_prob
                
                if len(sent) >= max_len:
                    sent.append("</s>")

            print( ' '.join(sent))
            print("句子的概率是：", -1/math.log(prob))

In [58]:
## 模型的建立
class LanguageModel():
    def __init__(self, tokens, n):
        self.tokens = tokens
        self.n = n
        self.vocab  = nltk.FreqDist(self.tokens)
        self.vocabulary_size = len(set(self.tokens)) 
        self.model = self.train_model()
        self.mask = list(reversed(list(product((0,1), repeat=n))))
        
    def train_model(self):
        """建立模型
        
        Returns:
            N-gram概率
        """
        if self.n == 1:
            num_tokens = len(self.tokens)
            # n_grams = nltk.ngrams(self.tokens, self.n)
            return { (unigram,): (count + 1) /(num_tokens + 1*self.vocabulary_size) for unigram, count in self.vocab.items() }
        else:
            n_grams = nltk.ngrams(self.tokens, self.n)
            n_vocab = nltk.FreqDist(n_grams)

            m_grams = nltk.ngrams(self.tokens, self.n-1)
            m_vocab = nltk.FreqDist(m_grams)
            def smoothed_count(n_gram, n_count):
                m_gram = n_gram[:-1]
                m_count = m_vocab[m_gram]
                return (n_count + 1) / (m_count + 1*self.vocabulary_size)
            return { n_gram: smoothed_count(n_gram, count) for n_gram, count in n_vocab.items() }

    def change_testtoken(self, ngram):
        """_summary_
            没有出现的字词换为<NUK>
        """
        mask = lambda ngram, bitmask: tuple((token if flag == 1 else "<UNK>" for token,flag in zip(ngram, bitmask)))

        ngram = (ngram,) if type(ngram) is str else ngram
        for possible_known in [mask(ngram, bitmask) for bitmask in self.mask]:
            if possible_known in self.model:
                return possible_known
            
    def perplexity(self, test_tokens):
        test_ngrams = nltk.ngrams(test_tokens, self.n)
        N = len(test_tokens)
        known_ngrams  = (self.change_testtoken(ngram) for ngram in test_ngrams)
        probabilities = [self.model[ngram] for ngram in known_ngrams]

        return math.exp((-1/N) * sum(map(math.log, probabilities)))
    
    def generate_sentences(self, num, min_len = 12, max_len = 24):
        def max_prob_word(prev, i, without=[]):
            blacklist = ["<UNK>"] + without
            candidates = ((ngram[-1],prob) for ngram,prob in self.model.items() if ngram[:-1]==prev)
            candidates = filter(lambda candidate: candidate[0] not in blacklist, candidates)
            word_list = sorted(candidates, key=lambda candidate: candidate[1], reverse=True)
            if len(word_list) == 0:
                return ("</s>", 1)
            else:
                return word_list[0 if prev != () and prev[-1] != "<s>" else i]
        
        for i in range(num):
            sent, prob = ["<s>"] * max(1, self.n-1), 1
            while sent[-1] != "</s>":
                prev = () if self.n == 1 else tuple(sent[-(self.n-1):])
                blacklist = sent + (["</s>"] if len(sent) < min_len else [])
                next_token, next_prob = max_prob_word(prev, i, without=blacklist)
                sent.append(next_token)
                prob *= next_prob
                
                if len(sent) >= max_len:
                    sent.append("</s>")

            yield ' '.join(sent), -math.log10(prob)
    

> 2) Report the perplexity of these 3 trained models on the testing dataset and explain your findings. 

In [16]:
# Your code goes to here

for n in [1,2,3]:
    print("-" * 50)
    train_tokens, train_sentences = token_text(json_objects_train, n, 1)
    test_tokens, test_sentences = token_text(json_objects_test, n, 1)
    print("Loading {}-gram model...".format(n))
    lm = LanguageModel(train_tokens, n)
    perplexity = lm.perplexity(test_tokens)
    print("Model perplexity: {:.3f}".format(perplexity))


--------------------------------------------------
Loading 1-gram model...
Model perplexity: 1172.520
--------------------------------------------------
Loading 2-gram model...
Model perplexity: 1906.412
--------------------------------------------------
Loading 3-gram model...
Model perplexity: 4956.581


可以看出模型并不是很好(perplexity稳步增加)

一般来说，应该2-gram的效果比1-gram要好，而3-gram的效果可能比2-gram略差（因为训练数据规模可能不够大）但还是应该比1-gram要好

说明可能模型某些超参数设置不是特别何理，我们将在第四部分讨论这个内容

> 3) Use each built model to generate five sentences and explain these generated patterns.


In [59]:
# Your code goes to here
## 对于1-gram
num_sent_gen = 5
print("-" * 25, "1-gram", "-"*25)
print("Loading {}-gram model...".format(1))
train_tokens, train_sentences = token_text(json_objects_train, 1, 1)
lm1 = LanguageModel(train_tokens, 1)
print("Generating sentences...")
for sentence, prob in lm1.generate_sentences(num_sent_gen):
    print("{} \n 句子的-log10(概率)是({:.5f})".format(sentence, prob))
        


------------------------- 1-gram -------------------------
Loading 1-gram model...
Generating sentences...
<s> the of and to in a was that as for with </s> 
 句子的-log10(概率)是(21.31607)
<s> of and to in a was that as for with is </s> 
 句子的-log10(概率)是(22.17082)
<s> and to in a was that as for with is his of by on The he from In an be at had which </s> 
 句子的-log10(概率)是(49.27788)
<s> to in a was that as for with is his by and on The he from In an be at had which it </s> 
 句子的-log10(概率)是(50.27780)
<s> in a was that as for with is his by on to The he from In an be at had which it are </s> 
 句子的-log10(概率)是(51.22885)


我们可以看到：

1. 生成的句子包含了一些常见的高频词汇，如"the"、"of"、"and"、"to"等，说明这些词汇在1-gram模型的语料库中出现频率很高。（一会验证）

2. 由于1-gram模型只考虑单个词的概率，因此生成的句子可能会缺乏语法结构和逻辑连贯性。例如，句子中可能会出现一些语法错误。

3. 生成的句子长度较短，通常由几个常见的词汇组成。

总的来说，1-gram模型生成的句子通常简单、短小，并且缺乏上下文的连贯性和语法结构。这是因为1-gram模型只考虑单个词的概率，而没有考虑词与词之间的关联性。因此，生成的句子可能会显得不够自然和连贯。

In [61]:
sorted(lm1.model.items(), key= lambda token:token[1], reverse=True)[:8]

[(('the',), 0.04979984708618549),
 (('<s>',), 0.04107336462852824),
 (('</s>',), 0.04107336462852824),
 (('of',), 0.03044493159681271),
 (('<UNK>',), 0.029501242546180168),
 (('and',), 0.027003277834880282),
 (('to',), 0.02205373745067218),
 (('in',), 0.01964199320339869)]

确实是一些介词连词代词概率大

In [62]:
## 对于2-gram
num_sent_gen = 5
print("-" * 25, "2-gram", "-"*25)
print("Loading {}-gram model...".format(2))
train_tokens, train_sentences = token_text(json_objects_train, 2, 1)
lm2 = LanguageModel(train_tokens, 2)
print("Generating sentences...")
for sentence, prob in lm2.generate_sentences(num_sent_gen):
    print("{} \n 句子的-log10(概率)是({:.5f})".format(sentence, prob))

------------------------- 2-gram -------------------------
Loading 2-gram model...
Generating sentences...
<s> The film was a new version of the first time in his own and that he had been found to be used for </s> 
 句子的-log10(概率)是(49.54975)
<s> In the first time of a new version was not be used to his own and that he had been found in which </s> 
 句子的-log10(概率)是(51.36339)
<s> He was a new version of the first time in his own and that he had been found to be used for its </s> 
 句子的-log10(概率)是(49.95609)
<s> It is a new version of the first time in his own and was not be used to have been found that he </s> 
 句子的-log10(概率)是(49.96390)
<s> This is a new version of the first time in his own and was not be used to have been found that he </s> 
 句子的-log10(概率)是(50.51037)


我们可以看出：
1. 生成的句子相比1-gram模型更具有上下文连贯性和语义合理性，因为2-gram模型考虑了相邻两个词之间的关联性。

2. 由于2-gram模型考虑了相邻两个词之间的关联性，生成的句子中会出现一些常见的词组和短语，开始出现除了介词连词助词以外的实意词。

3. 生成的句子长度适中，通常由几个常见的词组组成，但仍然受到局部上下文的限制。

4. 由于2-gram模型只考虑相邻两个词之间的关联性，生成的句子可能会出现一些语义不连贯或语法不对的地方。

总的来说，2-gram模型生成的句子相比1-gram模型更具有上下文连贯性和语义合理性，但仍然可能存在一些局部的语法错误或不连贯之处。这是因为2-gram模型考虑了相邻两个词之间的关联性，但仍然受到局部上下文的限制。

In [63]:
sorted(lm2.model.items(), key= lambda token:token[1], reverse=True)[:8]

[(('</s>', '<s>'), 0.6631479263025188),
 (('of', 'the'), 0.14265731617551342),
 (('in', 'the'), 0.11167441163243891),
 (('<UNK>', '</s>'), 0.1066721641132474),
 (('<s>', 'The'), 0.08030492264195706),
 (('to', 'the'), 0.06194222008432133),
 (('on', 'the'), 0.058464159779865334),
 (('<s>', 'In'), 0.05654950385368732)]

可以看出概率在前面的还是介词连词动词的组合

In [64]:
## 对于3-gram
num_sent_gen = 5
print("-" * 25, "3-gram", "-"*25)
print("Loading {}-gram model...".format(3))
train_tokens, train_sentences = token_text(json_objects_train, 3, 1)
lm3 = LanguageModel(train_tokens, 3)
print("Generating sentences...")
for sentence, prob in lm3.generate_sentences(num_sent_gen):
    print("{} \n 句子的-log10(概率)是({:.5f})".format(sentence, prob))

------------------------- 3-gram -------------------------
Loading 3-gram model...
Generating sentences...
<s> <s> The film was released on October 31, 2014. </s> 
 句子的-log10(概率)是(23.57530)
<s> <s> In the United States and Canada, it is not a single large room which was released on October 31, 2014. </s> 
 句子的-log10(概率)是(73.40615)
<s> <s> He was also a member of the most common cause with </s> 
 句子的-log10(概率)是(32.02132)
<s> <s> It is a major role in the United States and Canada, it had been an important part of his own work on </s> 
 句子的-log10(概率)是(72.01382)
<s> <s> This is the most common cause of death in a letter to his own work on "The Lord </s> 
 句子的-log10(概率)是(60.84203)


我们可以看到：
1. 3-gram有了合理的上下文关联：由于考虑了前两个单词的影响，生成的句子会更符合自然语言的语境，单词之间的关联性更强，句子更具有连贯性。
2. 生成的句子中常见短语和结构。
3. 句子有了更好的语法和语义：相比于2-gram模型，3-gram模型能够更好地捕捉到单词之间的语法和语义关系，生成的句子在语法和语义上更加合理。
4. 可以生成一个符合语法有语义的句子，如第一句话
   
总的来说，生成更好的句子的原因在于3-gram模型考虑了更多的上下文信息，能够更好地预测下一个单词出现的概率，从而生成更加合理的句子。通过考虑更多的上下文信息，3-gram模型能够更好地捕捉到单词之间的关联性，生成更接近自然语言的句子。

In [66]:
sorted(lm3.model.items(), key= lambda token:token[1], reverse=True)[:10]

[(('</s>', '<s>', '<s>'), 0.6631492265160873),
 (('<UNK>', '</s>', '<s>'), 0.20477678261473303),
 (('<s>', '<s>', 'The'), 0.08030492264195706),
 (('<s>', '<s>', 'In'), 0.05654950385368732),
 (('<s>', '<s>', 'He'), 0.029157051238062),
 (('<s>', '<s>', '<UNK>'), 0.02014975560309079),
 (('<s>', 'In', 'the'), 0.02012001236007695),
 (('<UNK>', '<UNK>', '<UNK>'), 0.019895673455842703),
 (('and', '<UNK>', '</s>'), 0.017530110318328518),
 (('as', 'well', 'as'), 0.01658324514486743)]

看起来未知的单词概率更大
也有as, well, as 这样的短语

> 4) 模型改良（不用laplace = 1 及增大UNK数量

In [40]:
for k in [2, 4, 6, 10, 20]:
    for n in [1,2,3]:
        print("-" * 50)
        train_tokens, train_sentences = token_text(json_objects_train, n, k)
        test_tokens, test_sentences = token_text(json_objects_test, n, k)
        # print("Loading {}-gram model...".format(n))
        lm = LanguageModel(train_tokens, n)
        perplexity = lm.perplexity(test_tokens)
        print("when freq <= {} turn into <UNK> , Loading {}-gram model...".format(k, n), "Model perplexity: {:.3f}".format(perplexity))

--------------------------------------------------


when freq <= 2 turn into <UNK> , Loading 1-gram model... Model perplexity: 833.608
--------------------------------------------------
when freq <= 2 turn into <UNK> , Loading 2-gram model... Model perplexity: 1160.163
--------------------------------------------------
when freq <= 2 turn into <UNK> , Loading 3-gram model... Model perplexity: 3260.697
--------------------------------------------------
when freq <= 4 turn into <UNK> , Loading 1-gram model... Model perplexity: 543.978
--------------------------------------------------
when freq <= 4 turn into <UNK> , Loading 2-gram model... Model perplexity: 629.916
--------------------------------------------------
when freq <= 4 turn into <UNK> , Loading 3-gram model... Model perplexity: 1891.847
--------------------------------------------------
when freq <= 6 turn into <UNK> , Loading 1-gram model... Model perplexity: 405.606
--------------------------------------------------
when freq <= 6 turn into <UNK> , Loading 2-gram model... Mo

可以看出，随着设置的UNK阈值的增大，模型的困惑度下降， 可能是由于以下原因：

1. 减少罕见词汇的干扰：更多的低频词被标记为UNK，这样模型就不用太过关注那些不常见的词汇，更专注于学习常见词汇的规律，从而降低困惑度。

2. 简化处理：减少词汇表的大小让模型更轻松地学习和泛化，不用处理太多生僻词汇，这有助于提升模型的表现。

3. 更好的适应性：增大UNK阈值让模型更多地把低频词汇看作同一类未知词，这有助于模型更好地适应处理新领域或不常见词汇，提升泛化能力。

总的来说，调整UNK阈值可以让模型更宽泛地处理文本，不被罕见词汇影响，表现得更好。

但是UNK阈值太高，可能会造成生成的句子不好，语言模型太宽泛

由于1-gram生成出来都是介词，在这里我们省略其生成，只看2-gram和3-gram

In [68]:
num_sent_test = 3
for k in [2, 4, 6, 10, 20]:
    for n in [2,3]:
        print("-" * 50)
        train_tokens, train_sentences = token_text(json_objects_train, n, k)
        test_tokens, test_sentences = token_text(json_objects_test, n, k)
        print("when freq <= {} turn into <UNK> , Loading {}-gram model...".format(k, n))
        lm = LanguageModel(train_tokens, n)
        for sentence, prob in lm.generate_sentences(num_sent_test):
            print("{} \n 句子的-log10(概率)是({:.5f})".format(sentence, prob))

--------------------------------------------------
when freq <= 2 turn into <UNK> , Loading 2-gram model...
<s> The film was a new version of the first time in his own and that he had been found to be used for </s> 
 句子的-log10(概率)是(46.79387)
<s> In the first time of a new version was not be used to his own and that he had been found in which </s> 
 句子的-log10(概率)是(48.60860)
<s> He was a new version of the first time in his own and that he had been found to be used for its </s> 
 句子的-log10(概率)是(47.21784)
--------------------------------------------------
when freq <= 2 turn into <UNK> , Loading 3-gram model...
<s> <s> The film was released on October 31, 2014. </s> 
 句子的-log10(概率)是(22.40032)
<s> <s> In the United States and Canada, it is not a single large room which was released on October 31, 2014. </s> 
 句子的-log10(概率)是(70.07187)
<s> <s> He was also a member of the most common cause with </s> 
 句子的-log10(概率)是(30.37858)
--------------------------------------------------
when freq <= 4 t

嘶可能生成的句子太少了 有点区分不出来

我们居中挑选了6

我们还可以更改laplace值，原代码中设置了1， 我们可以改成2，4， 0.1， 0.01 看看效果

In [77]:
## 模型的建立
class LanguageModel_2(object):
    def __init__(self, tokens, n, laplace):
        self.tokens = tokens
        self.n = n
        self.laplace = laplace
        self.vocab  = nltk.FreqDist(self.tokens)
        self.vocabulary_size = len(set(self.tokens)) 
        self.model = self.train_model()
        self.mask = list(reversed(list(product((0,1), repeat=n))))
        
    def train_model(self):
        """建立模型
        
        Returns:
            N-gram概率
        """
        if self.n == 1:
            num_tokens = len(self.tokens)
            # n_grams = nltk.ngrams(self.tokens, self.n)
            return { (unigram,): (count + self.laplace) /(num_tokens + self.laplace*self.vocabulary_size) for unigram, count in self.vocab.items() }
        else:
            n_grams = nltk.ngrams(self.tokens, self.n)
            n_vocab = nltk.FreqDist(n_grams)

            m_grams = nltk.ngrams(self.tokens, self.n-1)
            m_vocab = nltk.FreqDist(m_grams)
            def smoothed_count(n_gram, n_count):
                m_gram = n_gram[:-1]
                m_count = m_vocab[m_gram]
                return (n_count + self.laplace) / (m_count + self.laplace*self.vocabulary_size)
            return { n_gram: smoothed_count(n_gram, count) for n_gram, count in n_vocab.items() }

    def change_testtoken(self, ngram):
        """_summary_
            没有出现的字词换为<NUK>
        """
        mask = lambda ngram, bitmask: tuple((token if flag == 1 else "<UNK>" for token,flag in zip(ngram, bitmask)))

        ngram = (ngram,) if type(ngram) is str else ngram
        for possible_known in [mask(ngram, bitmask) for bitmask in self.mask]:
            if possible_known in self.model:
                return possible_known
            
    def perplexity(self, test_tokens):
        test_ngrams = nltk.ngrams(test_tokens, self.n)
        N = len(test_tokens)
        known_ngrams  = (self.change_testtoken(ngram) for ngram in test_ngrams)
        probabilities = [self.model[ngram] for ngram in known_ngrams]

        return math.exp((-1/N) * sum(map(math.log, probabilities)))
    
    def generate_sentences(self, num, min_len = 12, max_len = 24):
        def max_prob_word(prev, i, without=[]):
            blacklist = ["<UNK>"] + without
            candidates = ((ngram[-1],prob) for ngram,prob in self.model.items() if ngram[:-1]==prev)
            candidates = filter(lambda candidate: candidate[0] not in blacklist, candidates)
            word_list = sorted(candidates, key=lambda candidate: candidate[1], reverse=True)
            if len(word_list) == 0:
                return ("</s>", 1)
            else:
                return word_list[0 if prev != () and prev[-1] != "<s>" else i]
        
        for i in range(num):
            sent, prob = ["<s>"] * max(1, self.n-1), 1
            while sent[-1] != "</s>":
                prev = () if self.n == 1 else tuple(sent[-(self.n-1):])
                blacklist = sent + (["</s>"] if len(sent) < min_len else [])
                next_token, next_prob = max_prob_word(prev, i, without=blacklist)
                sent.append(next_token)
                prob *= next_prob
                
                if len(sent) >= max_len:
                    sent.append("</s>")

            yield ' '.join(sent), -math.log10(prob)
    

In [78]:
for k in [2, 4, 0.1, 0.01]:
    for n in [1,2,3]:
        print("-" * 50)
        train_tokens, train_sentences = token_text(json_objects_train, n, 6)
        test_tokens, test_sentences = token_text(json_objects_test, n, 6)
        # print("Loading {}-gram model...".format(n))
        lm = LanguageModel_2(train_tokens, n, laplace=k)
        perplexity = lm.perplexity(test_tokens)
        print("when laplace = {} , Loading {}-gram model...".format(k, n), "Model perplexity: {:.3f}".format(perplexity))

--------------------------------------------------
when laplace = 2 , Loading 1-gram model... Model perplexity: 407.701
--------------------------------------------------
when laplace = 2 , Loading 2-gram model... Model perplexity: 601.284
--------------------------------------------------
when laplace = 2 , Loading 3-gram model... Model perplexity: 2069.279
--------------------------------------------------
when laplace = 4 , Loading 1-gram model... Model perplexity: 411.936
--------------------------------------------------
when laplace = 4 , Loading 2-gram model... Model perplexity: 879.841
--------------------------------------------------
when laplace = 4 , Loading 3-gram model... Model perplexity: 3207.460
--------------------------------------------------
when laplace = 0.1 , Loading 1-gram model... Model perplexity: 403.737
--------------------------------------------------
when laplace = 0.1 , Loading 2-gram model... Model perplexity: 158.807
----------------------------------

可以发现laplace = 0.01效果最好

当 Laplace 参数较大（如4）时，模型的困惑度（perplexity）通常会增加，这意味着模型在对文本进行建模时变得更加困难，可能由于平滑过度导致模型过于简化。
当 Laplace 参数较小时（如0.01），模型的困惑度通常会减小，这意味着模型更好地适应了训练数据，但也可能导致过拟合。

可以生成句子康康效果：

In [80]:
for n in [1,2,3]:
    print("-" * 50)
    train_tokens, train_sentences = token_text(json_objects_train, n, 6)
    test_tokens, test_sentences = token_text(json_objects_test, n, 6)
    print("Loading {}-gram model...".format(n))
    lm = LanguageModel_2(train_tokens, n, laplace=0.01)
    for sentence, prob in lm.generate_sentences(5):
        print("{} \n 句子的-log10(概率)是({:.5f})".format(sentence, prob))

--------------------------------------------------
Loading 1-gram model...
<s> the of and to in a was that as for with </s> 
 句子的-log10(概率)是(21.20661)
<s> of and to in a was that as for with is </s> 
 句子的-log10(概率)是(22.06137)
<s> and to in a was that as for with is his of by on The he from In an be at had which </s> 
 句子的-log10(概率)是(49.06820)
<s> to in a was that as for with is his by and on The he from In an be at had which it </s> 
 句子的-log10(概率)是(50.06814)
<s> in a was that as for with is his by on to The he from In an be at had which it are </s> 
 句子的-log10(概率)是(51.01921)
--------------------------------------------------
Loading 2-gram model...
<s> The film was a new version of the first time in his own and that he had been found to be used for </s> 
 句子的-log10(概率)是(30.65861)
<s> In the first time of a new version was not be used to his own and that he had been found in which </s> 
 句子的-log10(概率)是(32.64052)
<s> He was a new version of the first time in his own and that he had been

## Task3 - Build NB/LR classifiers

> 1) Build a Naive Bayes classifier (with Laplace smoothing) and test your model on test dataset

数据处理直接引用LR的Feature 以及词袋，模型代码按照讲义编写

In [10]:
# Your code goes to here
import string
import math
import numpy as np
import json
### 数据处理
## 直接load LR 的feature使用
with open('X_train_dataset.json', 'r') as f:
    matrix_list1 = json.load(f)
X_train_dataset = np.array(matrix_list1)

with open('X_val_dataset.json', 'r') as f:
    matrix_list2 = json.load(f)
X_val_dataset = np.array(matrix_list2)

with open('X_test_dataset.json', 'r') as f:
    matrix_list3 = json.load(f)
X_test_dataset = np.array(matrix_list3)

## 加载标签
from sklearn import model_selection
X = [] # an element of X is represented as (title,text)
Y = [] # an element of Y represents the label of the corresponding X element
for i in range(1000):
    obj = json.loads(json_objects_train[i])
    title = obj.get("title")
    text = obj.get("text")
    label = obj.get("label")
    X.append((title, text))
    Y.append(label)

X_train, X_val, Y_train, Y_val = model_selection.train_test_split(X, Y, test_size=0.2, random_state=0)

Y_test = []
for i in range(100):
    obj = json.loads(json_objects_test[i])
    title = obj.get("title")
    text = obj.get("text")
    label = obj.get("label")
    Y_test.append(label)
    
## 合并数据
X_train_all = np.vstack((X_train_dataset, X_val_dataset))

Y_train_all = Y_train + Y_val

## 算出所有单词个数
V = len(X_train_all[0])

class_label = ['Film','Book','Politician','Writer','Food','Actor','Animal','Software','Artist','Disease']
class_label_dict = []

for i in range(10):
    all_num_vc = 0
    list_class = [0]*V
    for j in range(1000):
        if Y_train_all[j] == class_label[i]:
            all_num_vc +=  np.sum(X_train_all[j])
            list_class += X_train_all[j]
    class_label_dict.append([class_label[i], list_class, all_num_vc])

In [47]:
class NB_train_model():
    def __init__(self, dataset, laplace = 1):
        self.dataset = dataset
        self.laplace = laplace
    
    def every_class_prior(self, label, class_label, class_label_dict, V):
        """_summary_
            得到一个这个类别词汇的先验概率字典
        Args:
            label : 需要的类的名字
            class_label : 类别标签列表
            class_label_dict : 类别标签字典
            V : 词汇表大小

        Returns:
            word_prior_list : 每个单词先验概率列表
        """
        for i in range(10):
            if class_label[i] == label:
                N_c = class_label_dict[i][2]
                list_class_now = class_label_dict[i][1]
                word_prior_list = [0]*V
        
                for j in range(V):
                    word_prior_list[j] = (list_class_now[j] + self.laplace) / (N_c + V*self.laplace)
                    
                return word_prior_list
            
        # 如果没有找到匹配的类别标签，返回默认值
        return None

## train p(c)   
total = sum(train_classes_count.values())
train_class_prob = {key: value / total for key, value in train_classes_count.items()}
NB_model = NB_train_model(json_objects_train, laplace=1)

In [48]:
## 测试
import math
def test_on_model(model, test_dataset):
    correct_num = 0
    label_predict = []
    for i in range(100):
        model_label = 'None'
        model_prob = float("-inf")
        test_list = test_dataset[i]
        
        for class_name in train_classes_count.keys():
            model_prob_now = math.log2(train_class_prob[class_name])
            word_prior_list = model.every_class_prior(class_name, class_label, class_label_dict, V) ## 词汇概率字典
            
            for j in range(V):
                if word_prior_list[j] != 0:
                    model_prob_now += (math.log2(word_prior_list[j]))*test_list[j]
                
            ## 判断是否是最大的
            if model_prob_now > model_prob:
                model_prob = model_prob_now
                model_label = class_name
        
        ## 判断准确率
        if model_label == Y_test[i]:
            correct_num += 1
            print("correct", i, Y_test[i])
            label_predict.append(model_label)
        else:
            print("Wrong", i, "原来：",Y_test[i], "被分类为：",model_label)
            label_predict.append(model_label)
        
    return correct_num/100, label_predict

accuracy, label_predict = test_on_model(NB_model, X_test_dataset)
print(accuracy)

correct 0 Film
correct 1 Film
correct 2 Film
correct 3 Film
correct 4 Film
Wrong 5 原来： Film 被分类为： Politician
correct 6 Film
correct 7 Film
correct 8 Film
correct 9 Film
correct 10 Book
Wrong 11 原来： Book 被分类为： Writer
Wrong 12 原来： Book 被分类为： Politician
correct 13 Book
correct 14 Book
correct 15 Book
correct 16 Book
correct 17 Book
correct 18 Book
Wrong 19 原来： Book 被分类为： Writer
correct 20 Politician
correct 21 Politician
correct 22 Politician
correct 23 Politician
correct 24 Politician
correct 25 Politician
correct 26 Politician
correct 27 Politician
correct 28 Politician
correct 29 Politician
correct 30 Writer
correct 31 Writer
Wrong 32 原来： Writer 被分类为： Politician
correct 33 Writer
correct 34 Writer
correct 35 Writer
correct 36 Writer
correct 37 Writer
Wrong 38 原来： Writer 被分类为： Politician
correct 39 Writer
correct 40 Food
correct 41 Food
correct 42 Food
correct 43 Food
correct 44 Food
correct 45 Food
correct 46 Food
correct 47 Food
correct 48 Food
correct 49 Food
correct 50 Actor
correct

可以看到准确率为0.93

另：如果改变laplace值，可以看到效果差异不大

In [55]:

import math
def test_on_model(model, test_dataset):
    correct_num = 0
    # label_predict = []
    for i in range(100):
        model_label = 'None'
        model_prob = float("-inf")
        test_list = test_dataset[i]
        
        for class_name in train_classes_count.keys():
            model_prob_now = math.log2(train_class_prob[class_name])
            word_prior_list = model.every_class_prior(class_name, class_label, class_label_dict, V) ## 词汇概率字典
            
            for j in range(V):
                if word_prior_list[j] != 0:
                    model_prob_now += (math.log2(word_prior_list[j]))*test_list[j]
                
            ## 判断是否是最大的
            if model_prob_now > model_prob:
                model_prob = model_prob_now
                model_label = class_name
        
        # model_label = max(predict_prob.items(), key=lambda x:x[1])[0]
        ## 判断准确率
        if model_label == Y_test[i]:
            correct_num += 1
        #     print("correct", i, Y_test[i])
        #     label_predict.append(model_label)
        else:
            print("Wrong", i, "原来：",Y_test[i], "被分类为：",model_label)
            label_predict.append(model_label)
        
    return correct_num/100

for k in [2, 4, 0.5, 0.25]:
    NB_model = NB_train_model(json_objects_train, laplace=k)
    print("-"*15, "laplace为", k, "-"*15)
    print("被分错的")
    accuracy = test_on_model(NB_model, X_test_dataset)
    print(accuracy, "laplace =", k)

--------------- laplace为 2 ---------------
被分错的
Wrong 5 原来： Film 被分类为： Politician
Wrong 11 原来： Book 被分类为： Writer
Wrong 12 原来： Book 被分类为： Politician
Wrong 19 原来： Book 被分类为： Writer
Wrong 32 原来： Writer 被分类为： Politician
Wrong 38 原来： Writer 被分类为： Politician
Wrong 56 原来： Actor 被分类为： Writer
Wrong 59 原来： Actor 被分类为： Film
0.92 laplace = 2
--------------- laplace为 4 ---------------
被分错的
Wrong 5 原来： Film 被分类为： Politician
Wrong 11 原来： Book 被分类为： Writer
Wrong 12 原来： Book 被分类为： Politician
Wrong 16 原来： Book 被分类为： Writer
Wrong 19 原来： Book 被分类为： Writer
Wrong 32 原来： Writer 被分类为： Politician
Wrong 38 原来： Writer 被分类为： Politician
Wrong 50 原来： Actor 被分类为： Film
Wrong 51 原来： Actor 被分类为： Film
Wrong 52 原来： Actor 被分类为： Film
Wrong 53 原来： Actor 被分类为： Film
Wrong 54 原来： Actor 被分类为： Film
Wrong 55 原来： Actor 被分类为： Film
Wrong 56 原来： Actor 被分类为： Writer
Wrong 57 原来： Actor 被分类为： Film
Wrong 58 原来： Actor 被分类为： Film
Wrong 59 原来： Actor 被分类为： Film
Wrong 60 原来： Animal 被分类为： Disease
0.82 laplace = 4
--------------- laplace为 0.5 --

可以看到被分错的还是那几篇文章

> 2) Build a LR classifier. This question seems to be challenging. We did not directly provide features for samples. But just use your own method to build useful features. You may need to split the training dataset into train and validation so that some involved parameters can be tuned. 

数据处理代码参考老师的L03文件

In [102]:
# Your code goes to here
from sklearn import model_selection
import numpy as np
import nltk
import json
import torch
from torch.utils.data import Dataset, DataLoader
from torchvision.transforms import ToTensor
import torch.nn as nn
import torch.optim as optim
from sklearn.preprocessing import LabelEncoder

### 数据处理代码参考老师的L03文件
stopwords = ['a', 'about', 'above', 'across', 'after', 'afterwards', 'again', 'against', 'all', 'almost', 'alone',
             'along', 'already', 'also', 'although', 'always', 'am', 'among', 'amongst', 'amoungst', 'amount',
             'an', 'and', 'another', 'any', 'anyhow', 'anyone', 'anything', 'anyway', 'anywhere', 'are', 'around',
             'as', 'at', 'back', 'be', 'became', 'because', 'become', 'becomes', 'becoming', 'been', 'before',
             'beforehand', 'behind', 'being', 'below', 'beside', 'besides', 'between', 'beyond', 'bill', 'both',
             'bottom', 'but', 'by', 'call', 'can', 'cannot', 'cant', 'co', 'con', 'could', 'couldnt', 'cry', 'de',
             'describe', 'detail', 'did', 'do', 'does', 'doing', 'don', 'done', 'down', 'due', 'during', 'each', 'eg',
             'eight', 'either', 'eleven', 'else', 'elsewhere', 'empty', 'enough', 'etc', 'even', 'ever', 'every', 'everyone',
             'everything', 'everywhere', 'except', 'few', 'fifteen', 'fify', 'fill', 'find', 'fire', 'first', 'five', 'for',
             'former', 'formerly', 'forty', 'found', 'four', 'from', 'front', 'full', 'further', 'get', 'give', 'go', 'had',
             'has', 'hasnt', 'have', 'having', 'he', 'hence', 'her', 'here', 'hereafter', 'hereby', 'herein', 'hereupon',
             'hers', 'herself', 'him', 'himself', 'his', 'how', 'however', 'hundred', 'i', 'ie', 'if', 'in', 'inc', 'indeed',
             'interest', 'into', 'is', 'it', 'its', 'itself', 'just', 'keep', 'last', 'latter', 'latterly', 'least', 'less',
             'ltd', 'made', 'many', 'may', 'me', 'meanwhile', 'might', 'mill', 'mine', 'more', 'moreover', 'most', 'mostly',
             'move', 'much', 'must', 'my', 'myself', 'name', 'namely', 'neither', 'never', 'nevertheless', 'next', 'nine',
             'no', 'nobody', 'none', 'noone', 'nor', 'not', 'nothing', 'now', 'nowhere', 'of', 'off', 'often', 'on', 'once',
             'one', 'only', 'onto', 'or', 'other', 'others', 'otherwise', 'our', 'ours', 'ourselves', 'out', 'over', 'own',
             'part', 'per', 'perhaps', 'please', 'put', 'rather', 're', 's', 'same', 'see', 'seem', 'seemed', 'seeming',
             'seems', 'serious', 'several', 'she', 'should', 'show', 'side', 'since', 'sincere', 'six', 'sixty', 'so', 
             'some', 'somehow', 'someone', 'something', 'sometime', 'sometimes', 'somewhere', 'still', 'such', 'system',
             't', 'take', 'ten', 'than', 'that', 'the', 'their', 'theirs', 'them', 'themselves', 'then', 'thence', 'there',
             'thereafter', 'thereby', 'therefore', 'therein', 'thereupon', 'these', 'they', 'thickv', 'thin', 'third', 'this',
             'those', 'though', 'three', 'through', 'throughout', 'thru', 'thus', 'to', 'together', 'too', 'top', 'toward',
             'towards', 'twelve', 'twenty', 'two', 'un', 'under', 'until', 'up', 'upon', 'us', 'very', 'via', 'was', 'we',
             'well', 'were', 'what', 'whatever', 'when', 'whence', 'whenever', 'where', 'whereafter', 'whereas', 'whereby',
             'wherein', 'whereupon', 'wherever', 'whether', 'which', 'while', 'whither', 'who', 'whoever', 'whole', 'whom',
             'whose', 'why', 'will', 'with', 'within', 'without', 'would', 'yet', 'you', 'your', 'yours', 'yourself',
             'yourselves']

## 分出train 和 val
X = [] # an element of X is represented as (title,text)
Y = [] # an element of Y represents the label of the corresponding X element
for i in range(1000):
    obj = json.loads(json_objects_train[i])
    title = obj.get("title")
    text = obj.get("text")
    label = obj.get("label")
    X.append((title, text))
    Y.append(label)

X_train, X_val, Y_train, Y_val = model_selection.train_test_split(X, Y, test_size=0.2, random_state=0)

## 选出所有feature
feature_word = []
for i in range(len(X_train)):
    text = X_train[i][1]
    tt = nltk.word_tokenize(text)
    tokens = [word for word in (word.strip(string.punctuation).lower() for word in tt) if word != '']
    # tokenized_word.extend(tokens)
    feature_word.extend(tokens)
vocab = nltk.FreqDist(feature_word)
feature_word = [token for token in feature_word if vocab[token] > 5 ] ## 限制大于5的频率才算
feature_word = [word for word in feature_word if word not in stopwords]
# print(len(set(feature_word))) 
# 37507
feature = list(set(feature_word))

# 把train_dataset 和 Val_dataset 变成向量（后边变成tensor）
X_train_dataset = np.zeros((len(X_train),len(feature)))
for i in range(len(X_train)):
    text = X_train[i][1]
    tt = nltk.word_tokenize(text)
    tokens = [word for word in (word.strip(string.punctuation).lower() for word in tt) if word != '']
    vocab = nltk.FreqDist(tokens)
    for word in vocab:
        if word in feature:
            X_train_dataset[i][feature.index(word)] = vocab[word]
            
# ## 保存一下 跑一次要好久
# matrix_list = X_train_dataset.tolist()
# with open('X_train_dataset.json', 'w') as f:
#     json.dump(matrix_list, f)

# ## 重新加载

# with open('X_train_dataset.json', 'r') as f:
#     matrix_list = json.load(f)

# # 将数据转换为矩阵
# matrix = np.array(matrix_list)

X_val_dataset = np.zeros((len(X_val),len(feature)))
for i in range(len(X_val)):
    text = X_val[i][1]
    tt = nltk.word_tokenize(text)
    tokens = [word for word in (word.strip(string.punctuation).lower() for word in tt) if word != '']
    vocab = nltk.FreqDist(tokens)
    for word in vocab:
        if word in feature:
            X_val_dataset[i][feature.index(word)] = vocab[word]
            
# matrix_list2 = X_val_dataset.tolist()
# with open('X_val_dataset.json', 'w') as f:
#     json.dump(matrix_list2, f)

## 转换为tensor形式
X_train_dataset_tensor = torch.from_numpy(X_train_dataset)
X_val_dataset_tensor = torch.from_numpy(X_val_dataset)
tensor_list_train = torch.cat([X_train_dataset_tensor[i, :].unsqueeze(0) for i in range(800)], dim=0)
tensor_list_val = torch.cat([X_val_dataset_tensor[i, :].unsqueeze(0) for i in range(200)],dim = 0)

## 把标签换成整数索引
class_label = ['Film','Book','Politician','Writer','Food','Actor','Animal','Software','Artist','Disease']
int_label = [5,3,7,9,6,0,1,8,2,4]
label_encoder = LabelEncoder()
Y_train_indices = label_encoder.fit_transform(Y_train)
Y_val_indices = label_encoder.fit_transform(Y_val)
# ## 转换回来
# original_labels = label_encoder.inverse_transform(Y_val_indices)

## 传入数据
class Dataset(Dataset):
    def __init__(self, data, labels):
        self.data = data
        self.labels = labels

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]

train_dataset = Dataset(tensor_list_train, Y_train_indices)
val_dataset = Dataset(tensor_list_val, Y_val_indices)

## 超参数
batch_size = 100
lr_rate = 0.01
num_epochs = 2
input_size = 37507
output_size = 10

train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=50, shuffle=False)


## 构建模型
class LogisticRegression(nn.Module):
    def __init__(self, input_size, output_size):
        super(LogisticRegression, self).__init__()
        self.linear = nn.Linear(input_size, output_size)

    def forward(self, x):
        return self.linear(x)

model = LogisticRegression(input_size, output_size)
model = model.to(dtype=torch.float64)
# 损失函数和优化器
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=lr_rate)

for epoch in range(num_epochs):
    for data in train_loader:
        imgs, targets = data
        output = model(imgs)
        loss = criterion(output, targets)

        # 反向传播和优化
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        # 打印epoch损失
    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item()}')

# 保留参数
# torch.save(model.state_dict(), 'model.pth')

# 在验证集上评估
correct = 0
total = 0
with torch.no_grad():
    for inputs, labels in val_loader:
        outputs = model(inputs)
        _, predicted = torch.max(outputs, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

accuracy = correct / total
print("Validation Accuracy:", accuracy)
# 算出来0.96

> 3) Report Micro-F1 score and Macro-F1 score for these classifiers on testing dataset explain our results.

In [39]:
# Your code goes to here
from sklearn.metrics import precision_recall_fscore_support
## 对于NB模型
print("-"*15, "NB model", "-"*15)
Y1 = []
for i in range(100):
    obj = json.loads(json_objects_test[i])
    title = obj.get("title")
    text = obj.get("text")
    label = obj.get("label")
    Y1.append(label)
    
# 计算每个类精确度、召回率和 F1 分数
precision, recall, f1, _ = precision_recall_fscore_support(Y1, label_predict, average=None)

# 计算 Micro-F1 分数
micro_precision, micro_recall, micro_f1, _ = precision_recall_fscore_support(Y1, label_predict, average='micro')

macro_f1 = np.mean(f1)

for i in range(output_size):
    print(f'Class {class_label[i]}: Precision={precision[i]}, Recall={recall[i]}, F1={f1[i]}')

print(f'Micro-F1 Score of LR model: {micro_f1}')
print(f'Macro-F1 Score of LR model: {macro_f1}')


--------------- NB model ---------------
Class Film: Precision=1.0, Recall=0.9, F1=0.9473684210526316
Class Book: Precision=1.0, Recall=1.0, F1=1.0
Class Politician: Precision=1.0, Recall=1.0, F1=1.0
Class Writer: Precision=1.0, Recall=0.7, F1=0.8235294117647058
Class Food: Precision=1.0, Recall=1.0, F1=1.0
Class Actor: Precision=1.0, Recall=0.9, F1=0.9473684210526316
Class Animal: Precision=1.0, Recall=1.0, F1=1.0
Class Software: Precision=0.7142857142857143, Recall=1.0, F1=0.8333333333333333
Class Artist: Precision=1.0, Recall=1.0, F1=1.0
Class Disease: Precision=0.7272727272727273, Recall=0.8, F1=0.761904761904762
Micro-F1 Score of LR model: 0.93
Macro-F1 Score of LR model: 0.9313504349108065


### 对于NB模型：
1. 对于部分类别（Class Book, Politician, Food, Animal, Artist），模型的精确度、召回率和 F1 分数都达到了 1.0，这意味着模型在这些类别上的预测非常准确，并且没有漏掉任何样本。
   
2. 在其他几个类别（Class Film, Writer, Actor, Software, Disease）上，模型的性能稍有下降。比如，在 Class Film, Writer, Actor 上，模型的召回率较低，表明模型未能捕获到所有正类样本。在 Class Software 上，模型的精确度稍低，表明模型在该类别上可能存在一些误报，而在 Class Disease 上，模型的精确度稍低，召回率也稍低，表明模型在该类别上的性能较差。
   
3. 总体而言，Micro-F1 分数为 0.93，Macro-F1 分数为 0.931。Micro-F1 分数考虑了所有类别，并给予每个类别相同的权重。在这个结果中，Micro-F1 分数为0.93，表明模型在整体上具有较好的性能。Macro-F1 分数是所有类别 F1 分数的平均值，它平等对待每个类别。在这个结果中，Macro-F1 分数为0.931，说明模型在每个类别上的较为性能均衡，没有某个类别的性能明显偏低。

In [34]:
from sklearn.metrics import precision_recall_fscore_support
import torch
from torch.utils.data import Dataset, DataLoader
from torchvision.transforms import ToTensor

### 对于LR模型
print("-"*15, "LR model", "-"*15)

# 打印模型参数
# for name, param in model.named_parameters():
#     print(f'Parameter name: {name}, Size: {param.size()}, Values: {param}')
    
## 加载测试数据
import numpy as np
X1 = []
Y1 = []
for i in range(100):
    obj = json.loads(json_objects_test[i])
    title = obj.get("title")
    text = obj.get("text")
    label = obj.get("label")
    X1.append((title, text))
    Y1.append(label)

# 从 JSON 文件中加载数据
with open('X_test_dataset.json', 'r') as f:
    matrix_list = json.load(f)

# 将数据转换为矩阵
X_test_dataset = np.array(matrix_list)

from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
class Dataset(Dataset):
    def __init__(self, data, labels):
        self.data = data
        self.labels = labels

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        return self.data[idx], self.labels[idx]

## 处理数据
X_test_dataset_tensor = torch.from_numpy(X_test_dataset)
tensor_list_test = torch.cat([X_test_dataset_tensor[i, :].unsqueeze(0) for i in range(100)],dim = 0)
Y_test_indices = label_encoder.fit_transform(Y1)
test_dataset = Dataset(tensor_list_test, Y_test_indices)
test_loader = DataLoader(test_dataset, batch_size=50, shuffle=False)

## 准备模型和数据加载
# 准备数据
input_size = 37507
output_size = 10
model = LogisticRegression(input_size, output_size)
model.load_state_dict(torch.load('model.pth'))

## 计算
model.eval()
model = model.to(dtype=torch.float64)

all_predictions = []
all_labels = []


with torch.no_grad():
    for data in test_loader:
        inputs, label = data
        outputs = model(inputs)
        _, predicted = torch.max(outputs, 1)
        all_predictions.extend(predicted.numpy())
        all_labels.extend(label.numpy())

# 计算每个类精确度、召回率和 F1 分数
precision, recall, f1, _ = precision_recall_fscore_support(all_labels, all_predictions, average=None)

# 计算 Micro-F1 分数
micro_precision, micro_recall, micro_f1, _ = precision_recall_fscore_support(all_labels, all_predictions, average='micro')

macro_f1 = np.mean(f1)

for i in range(output_size):
    print(f'Class {i}: Precision={precision[i]}, Recall={recall[i]}, F1={f1[i]}')

print(f'Micro-F1 Score of LR model: {micro_f1}')
print(f'Macro-F1 Score of LR model: {macro_f1}')

--------------- LR model ---------------
Class 0: Precision=1.0, Recall=1.0, F1=1.0
Class 1: Precision=1.0, Recall=1.0, F1=1.0
Class 2: Precision=1.0, Recall=1.0, F1=1.0
Class 3: Precision=0.8, Recall=0.8, F1=0.8000000000000002
Class 4: Precision=1.0, Recall=1.0, F1=1.0
Class 5: Precision=1.0, Recall=0.9, F1=0.9473684210526316
Class 6: Precision=1.0, Recall=1.0, F1=1.0
Class 7: Precision=0.9090909090909091, Recall=1.0, F1=0.9523809523809523
Class 8: Precision=0.9090909090909091, Recall=1.0, F1=0.9523809523809523
Class 9: Precision=0.8888888888888888, Recall=0.8, F1=0.8421052631578948
Micro-F1 Score of LR model: 0.9500000000000001
Macro-F1 Score of LR model: 0.9494235588972432


C:\Users\Kim\AppData\Local\Temp\ipykernel_22076\2623391384.py:18: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
  return F.softmax(self.linear(x))


### 对于LR模型：
1. 对于大多数类别（Class 0, 1, 2, 4, 6），模型的精确度、召回率和 F1 分数都达到了 1.0，这意味着模型在这些类别上的预测非常准确，并且没有漏掉任何样本。
   
2. 在其他几个类别（Class 3, 5, 7, 8, 9）上，模型的性能稍有下降。比如，在 Class 3, 5, 9 上，模型的召回率较低，表明模型未能捕获到所有正类样本。在 Class 1, 3, 7, 8 上，模型的精确度稍低，表明模型在该类别上可能存在一些误报，而在 Class 9 上，模型的精确度稍低，召回率也稍低，表明模型在该类别上的性能较差。
   
3. 总体而言，Micro-F1 分数为 0.95，Macro-F1 分数为 0.9494。Micro-F1 分数考虑了所有类别，并给予每个类别相同的权重。在这个结果中，Micro-F1 分数为0.95，表明模型在整体上具有较好的性能。Macro-F1 分数是所有类别 F1 分数的平均值，它平等对待每个类别。在这个结果中，Macro-F1 分数为0.9494，说明模型在每个类别上的较为性能均衡，没有某个类别的性能明显偏低。

总的来说，我建立的LR模型效果比NB模型要好